### PR TITLE
Refactor `GenerateUserCerts` to create App Sessions

### PR DESCRIPTION
**Pull request by Joerger**
Implement the changes to `GenerateUserCerts` described in [RFD 169](https://github.com/gravitational/teleport/blob/e23afa625a8357e7dd7ea7e7c848fd13742eeb40/rfd/0169-app-mfa-sessions.md#rpc-generateusercerts)

```md
#### `rpc GenerateUserCerts`

Currently, `rpc CreateAppSession` is used by direct clients to create an App
Session, retrieve it's session ID, and request local app certs linked to the
session ID. In order for direct clients to mark both the App Session cert and
local app cert as MFA verified, the user would need to perform two MFA checks
and two roundtrips.

This should be avoided by allowing clients to call `rpc GenerateUserCerts` to
generate both the local app cert and App Session with a single MFA response.
The client only needs the local app cert after all.
```

Backwards compatibility between v16 clients and v15 auth servers is handled through a best-effort version check to determine whether the client should create an app session separately. This replaces one roundtrip with another (Ping) but can be removed in v17.

Based off https://github.com/gravitational/teleport/pull/39742

### CODE REVIEW

File: lib/auth/sessions.go [hunk 1]
```diff
@@ -388,3 +463,44 @@ func (a *Server) CreateSAMLIdPSession(ctx context.Context, req types.CreateSAMLI
...
+	CreateAppSession(ctx context.Context, req *proto.CreateAppSessionRequest) (types.WebSession, error)
+	AppSessionWatcher
+}
+
+// TryCreateAppSessionForClientCertV15 creates an app session if the auth
+// server is pre-v16 and returns the app session ID. This app session ID
+// is needed for user app certs requests before v16.
+// TODO (Joerger): DELETE IN v17.0.0
+func TryCreateAppSessionForClientCertV15(ctx context.Context, client CreateAppSessionForV15Client, username string, routeToApp proto.RouteToApp) (string, error) {
+	pingResp, err := client.Ping(ctx)
```
Comment (zmb3) [inline lib/auth/sessions.go:478]: Do we need the extra round trip? Can we use `metadata.ClientVersionFromContext()` here?

Comment (Joerger) [reply lib/auth/sessions.go:478]: This is actually called by the client to check the server version, so we need the roundtrip.

Comment (greedy52) [reply lib/auth/sessions.go:638]: both `tc` and `tbot` have some cache over ping response. Ideally we can use those, but I see it's quite a hassle to use them here for a temporary change.

**Review by greedy52 — APPROVED** [review]

File: lib/auth/sessions.go [hunk 2]
```diff
@@ -255,19 +275,66 @@ func (a *Server) CreateAppSession(ctx context.Context, req *proto.CreateAppSessi
...
+		return nil, trace.Wrap(err)
+	}
+	checker, err := services.NewAccessChecker(&services.AccessInfo{
+		Roles:              user.GetRoles(),
+		Traits:             user.GetTraits(),
+		AllowedResourceIDs: req.RequestedResourceIDs,
+	}, clusterName.GetClusterName(), a)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
```
[the remark this answers is not in the document]
Comment (Joerger) [reply lib/auth/sessions.go:323]: The ttl can be variable depending on the caller - `GenerateUserCerts` provides the TTL requested in the generate certs request.

File: lib/auth/sessions.go [hunk 3]
```diff
@@ -255,19 +275,66 @@ func (a *Server) CreateAppSession(ctx context.Context, req *proto.CreateAppSessi
...
 		user:           user,
-		loginIP:        identity.LoginIP,
+		loginIP:        req.LoginIP,
 		publicKey:      publicKey,
 		checker:        checker,
-		ttl:            ttl,
-		traits:         traits,
-		activeRequests: services.RequestIDs{AccessRequests: identity.ActiveRequests},
+		ttl:            req.SessionTTL,
+		traits:         req.Traits,
```
[the remark this answers is not in the document]
Comment (Joerger) [reply lib/auth/sessions.go:336]: Yeah it seems like we may have some redundant ways of doing the same thing now, should we be moving towards using the user state when possible instead?

File: lib/auth/sessions.go [hunk 4]
```diff
@@ -255,19 +275,66 @@ func (a *Server) CreateAppSession(ctx context.Context, req *proto.CreateAppSessi
...
+
+	clusterName, err := a.GetClusterName()
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	checker, err := services.NewAccessChecker(&services.AccessInfo{
+		Roles:              user.GetRoles(),
+		Traits:             user.GetTraits(),
+		AllowedResourceIDs: req.RequestedResourceIDs,
+	}, clusterName.GetClusterName(), a)
```
[the remark this answers is not in the document]
Comment (Joerger) [reply lib/auth/sessions.go]: I tried applying this, but actually I don't think we want this - we want to get the traits and roles from the user's current certificate to include roles and traits from access requests.

[the remark this answers is not in the document]
Comment (greedy52) [reply lib/auth/sessions.go]: makes sense 👍

**Comment by Joerger** [comment]
@zmb3 @strideynet friendly ping to review

**Review by strideynet — APPROVED** [review]
Apologies - got through most of reviewing this and forgot to finish my review. Looks glad - I'm glad this is a single RPC now.

**Review by zmb3 — APPROVED** [review]

**Pull request merged**